impl send sync to world to use it with specs

thiolliere opened the pull request:
This PR is not ready yet. just to show you what I am working on

this implementation allow me to use nphysics with specs. I do that in https://github.com/thiolliere/pepe
I have `nphysics::world::World` as a resource and `RigidBody(usize)` component with method to get reference to the actual rigid body with borrowing the nphysics world resource.

(examples have not been updated)

The whole point was not to use Rc<RefCell<_>> for rigid bodies, sensor
and joint and instead use a storage and an index to handle them.

The storage must be efficient index_vector is my implementation:
it uses a vector to store the data and never move the position of the
elements in the vector. So the index of the vector always point to the
same element.
We use an offset in order to make sensor and rigid_bodies index
different.

still issue:
* fitler data is difficult to use as we can't send a tuple of
  reference to storage because it would imply a use of lifetime that the
  compiler can't handle (or it's me). I wanted `FilterData = (&'a
  RigidBodyStorage, &'a SensorStorage) but I couldn't impl it.

  I think the filter things could be different like instead of register
  callbacks in collision world. callbacks would be stored in world and
  collision_world.perform_broad_phase would have a unique callback
  argument. By doing so all the reference lifetimes would be in the in the
  function and so it should be OK.

  the only drawback of this is that it is to the world to ensure to call
  `broad_phase.deferred_recompute_all_proximities();`
* use my fork of ncollide

still todo:
* documentation
* maybe wrap sensor, joint and rigid_body index in a struct(usize) so user
  can't missuse them.
* API
* also I don't know about performance

dkushner commented:
I find this interesting, as I am also looking to integrate nphysics with specs.

@sebcrozet: What is the general opinion on this PR or a similar effort?

thiolliere commented:
This is usable but there is two things that are not very handy:

I use the following body component:
```rust
// Rigid body handle and whereas it has been deleted
pub struct PhysicBody {
    handle: usize,
    removed: bool,
}

impl ::specs::Component for PhysicBody {
    type Storage = ::specs::VecStorage<Self>;
}

impl Drop for PhysicBody {
    fn drop(&mut self) {
        if !self.removed {
            // debug_assert!(eprintln!("physic body hasn't been removed from physic world") == ());
        }
    }
}

impl PhysicBody {
    pub fn entity(body: &::nphysics::object::RigidBody<f32>) -> ::specs::Entity {

        let entity = body.user_data().unwrap();
        let entity = unsafe { ::std::mem::transmute::<&Box<_>, &Box<Any>>(entity) };
        entity.downcast_ref::<::specs::Entity>().unwrap().clone()
    }

    pub fn add<'a>(
        entity: ::specs::Entity,
        mut body: ::nphysics::object::RigidBody<f32>,
        bodies: &mut ::specs::WriteStorage<'a, ::component::PhysicBody>,
        physic_world: &mut ::specs::FetchMut<'a, ::resource::PhysicWorld>,
    ) {
        body.set_user_data(Some(Box::new(entity)));
        let bodyhandle = physic_world.add_rigid_body(body);
        bodies.insert(
            entity,
            PhysicBody {
                handle: bodyhandle,
                removed: false,
            },
        );
    }

    #[inline]
    pub fn ball_in_socket(
        &mut self,
        physic_world: &mut ::resource::PhysicWorld,
        position: ::na::Point3<f32>,
    ) {
        physic_world.add_ball_in_socket(joint::BallInSocket::new(
            joint::Anchor::new(None, position),
            joint::Anchor::new(
                Some(self.handle),
                ::na::Point3::new(0.0, 0.0, 0.0),
            ),
        ));
    }

    #[inline]
    pub fn get<'a>(
        &'a self,
        physic_world: &'a ::resource::PhysicWorld,
    ) -> &'a ::nphysics::object::RigidBody<f32> {
        physic_world.rigid_body(self.handle)
    }

    #[inline]
    pub fn get_mut<'a>(
        &'a mut self,
        physic_world: &'a mut ::resource::PhysicWorld,
    ) -> &'a mut ::nphysics::object::RigidBody<f32> {
        physic_world.mut_rigid_body(self.handle)
    }

    pub fn remove(&mut self, physic_world: &mut ::resource::PhysicWorld) {
        if self.removed {
            panic!("physic body already removed from physic world");
        }
        physic_world.remove_rigid_body(self.handle);
        self.removed = true;
    }
}
```

* 1- each time I really want to access body component I have to use the physic world resource borrow it like that.
```rust
let body = body.get_mut(&mut physic_world);
```

* 2- but worse when I want to remove an entity I have to explicitly remove it from physic_world resource like that:
```rust
if let Some(body) = bodies.get_mut(entity) {
    body.remove(&mut physic_world);
}
entities.delete(entity).unwrap();
```

To solve 1 we can externalise storage of bodies and sensors and joints. and give storages when you want to update world and do things.

Solve 2 is more difficult. I don't know how I can access deleted and created entities without having to store all entites with body component and a system that at each specs_world.maintain find created and deleted by comparing with the copy.
Accessing created and deleted entities can be wanted by other there may be a feature that is missing from specs here.
It would be cool to have a callback that is called with resource for each entity deleted and created.

thiolliere commented:
@dkushner sebcrozet is making a full rewrite with Send Sync world #101 and you can see his plan on https://www.patreon.com/sebcrozet

also to solve 2 a new storage appeared in specs master: trackedstorage that track insert modified, removed entities on storage. This can be used to automatically remove deleted entities from physic world.

There is so an inconsistence between entities in specs world and entities in physics world: deleted entities remain in physic world until next trackedstorage maintain.

sebcrozet commented:
Thanks a lot for all your work here @thiolliere!
This gave ideas for the redesign of nphysics. I'm closing this since the newest vesion of nphysics defines a `Send + Sync` world.
